encoding/protobuf/encodingIntegration_test.go

```diff
@@ -0,0 +1,135 @@
+
+package protobuf_test
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/gogo/protobuf/types"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/yarpc"
+	"go.uber.org/yarpc/encoding/protobuf"
+	"go.uber.org/yarpc/encoding/protobuf/internal/testpb"
+	"go.uber.org/yarpc/yarpcerrors"
+	"go.uber.org/yarpc/yarpctest"
+)
+
+type testStructure struct {
+	name   string
+	req    *yarpctest.Call
+	expReq map[string]string
+}
+
+var allTests map[string]testStructure
+
+func validateReq(testname string, ctx context.Context) (bool, string) {
+	test := allTests[testname]
+	call := yarpc.CallFromContext(ctx)
+	for name, value := range test.expReq {
+		switch name {
+		case "CallerProcedure":
+			if call.CallerProcedure() != value {
+				err := "CallerProcedure '" + call.CallerProcedure() + "' does match with expected value '" + value + "'"
+				return false, err
+			}
+		case "Procedure":
+			if call.Procedure() != value {
+				err := "Procedure '" + call.Procedure() + "' does match with expected value '" + value + "'"
+				return false, err
+			}
+		}
+	}
+	//fmt.Println("Entered in validateReq :  ", testname)
+	return true, ""
+}
+
+func runTest(t *testing.T, test testStructure, client testpb.TestYARPCClient) {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	ctx = yarpctest.ContextWithCall(ctx, test.req)
+	defer cancel()
+
+	_, err := client.Unary(ctx, &testpb.TestMessage{Value: test.name})
+	require.NoError(t, err, "unexpected call error")
+}
+
+func TestProtobufEncoding(t *testing.T) {
+	client, _, _, _, cleanup := initClientAndServer(t, &encodingIntegrationTestServer{})
+
+	defer cleanup()
+
+	tests := []testStructure{
+		{
+			name: "test1",
+			req: &yarpctest.Call{
+				Procedure: "ABC1",
+			},
+			expReq: map[string]string{
+				"CallerProcedure": "ABC1",
+				"Procedure":       "uber.yarpc.encoding.protobuf.Test::Unary",
+			},
+		},
+		{
+			name: "test2",
+			req:  &yarpctest.Call{},
+			expReq: map[string]string{
+				"CallerProcedure": "",
+				"Procedure":       "uber.yarpc.encoding.protobuf.Test::Unary",
+			},
+		},
+	}
+	allTests = make(map[string]testStructure)
+	for _, test := range tests {
+		allTests[test.name] = test
+		runTest(t, test, client)
+	}
+
+}
+
+type encodingIntegrationTestServer struct{}
+
+func (encodingIntegrationTestServer) Unary(ctx context.Context, msg *testpb.TestMessage) (*testpb.TestMessage, error) {
+	ok, err := validateReq(msg.Value, ctx)
+	if ok == true {
+		return &testpb.TestMessage{Value: msg.Value}, nil
+	}
+
+	details := []proto.Message{
+		&types.StringValue{Value: err},
+	}
+
+	return nil, protobuf.NewError(yarpcerrors.CodeInvalidArgument, err, protobuf.WithErrorDetails(details...))
+}
+
+func (encodingIntegrationTestServer) Duplex(stream testpb.TestServiceDuplexYARPCServer) error {
+	for {
+		msg, err := stream.Recv()
+		if err != nil {
+			return err
+		}
+		err = stream.Send(msg)
+		if err != nil {
+			return err
+		}
+	}
+}
```

Comment:
Thanks for adding this tests, there is clearly a lack of e2e regarding the header propagated in this package.

Few suggestions here:

Can we add this test in the file `outbound_test.go`, this file already contain some e2e tests about outbound request. We could have a new test `TestOutboundHeaders`.

Is there any chance to reduce the complexity of the test by using `yarpctest`? We could have something like
```
			out := trans.NewOutbound(nil, yarpctest.OutboundCallOverride(
				yarpctest.OutboundCallable(func(ctx context.Context, req *transport.Request) (*transport.Response, error) {
					return &transport.Response{Body: ioutil.NopCloser(ctx.Headers())}, nil
				}),
			))
```
The test will then check the response and the header returned. There are multiple benefits for this:
- It removes a lot of code :) - no need of refactoring things in `observability_test.go`. The PR will become way smaller and easy to review ;)
- All logic to the test is contained into the test, there is no need to understand how a temp server is working and what is returning. No need of maintaining global vars as well. Easy to maintain!

Thoughts?

Reply:
Thank you for reviewing.  I follow the DRY principle very religiously that's the reason I have refactored this test code. However, as discussed offline wrt to simplify the test by using the above construct. So I will accept this comment, and will make the changes accordingly.